common/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,847 @@
+            norms_rhs_sh, static_cast<size_type>(nrhs), 1, nrhs};
+
+        const gko::batch_dense::BatchEntry<real_type> res_norms_shared_entry{
+            norms_res_sh, static_cast<size_type>(nrhs), 1, nrhs};
+
+        const gko::batch_dense::BatchEntry<real_type>
+            res_norms_temp_shared_entry{norms_res_temp_sh,
+                                        static_cast<size_type>(nrhs), 1, nrhs};
+
+        // generate preconditioner
+        prec_shared.generate(A_global_entry, prec_work_sh);
+
+
+        // initialization
+        // rho_old = 1, omega_old = 1, alpha = 1
+        // compute b norms
+        // copy x from global to shared memory
+        // r = b - A*x
+        // compute residual norms
+        // r_hat = r
+        // p = 0
+        // v = 0
+        initialize(rho_old_shared_entry, omega_old_shared_entry,
+                   alpha_shared_entry, A_global_entry, b_global_entry,
+                   gko::batch::to_const(x_global_entry), x_shared_entry,
+                   r_shared_entry, r_hat_shared_entry, p_shared_entry,
+                   v_shared_entry, rhs_norms_shared_entry,
+                   res_norms_shared_entry);
+
+
+        __syncthreads();
+
+
+        // stopping criterion object
+        StopType stop(nrhs, max_iter, abs_tol, rel_tol,
+                      static_cast<stop::tolerance>(tol_type),
+                      rhs_norms_shared_entry.values, converged);
+
+        int iter = -1;
+
+
+        while (1) {
+            ++iter;
+
+            bool all_converged =
+                stop.check_converged(iter, res_norms_shared_entry.values,
+                                     {NULL, 0, 0, 0}, converged);
+
+
+            logger.log_iteration(ibatch, iter, res_norms_shared_entry.values,
+                                 converged);
+            if (all_converged) {
+                break;
+            }
+
+
+            // rho_new =  < r_hat , r > = (r_hat)' * (r)
+            compute_dot_product<ValueType>(
+                gko::batch::to_const(r_hat_shared_entry),
+                gko::batch::to_const(r_shared_entry), rho_new_shared_entry);
+
+
+            __syncthreads();
+
+            // beta = (rho_new / rho_old)*(alpha / omega_old)
+            compute_beta(gko::batch::to_const(rho_new_shared_entry),
+                         gko::batch::to_const(rho_old_shared_entry),
+                         gko::batch::to_const(alpha_shared_entry),
+                         gko::batch::to_const(omega_old_shared_entry),
+                         beta_shared_entry, converged);
+
+            __syncthreads();
+
+
+            // p = r + beta*(p - omega_old * v)
+            update_p(gko::batch::to_const(r_shared_entry),
+                     gko::batch::to_const(beta_shared_entry),
+                     gko::batch::to_const(omega_old_shared_entry),
+                     gko::batch::to_const(v_shared_entry), p_shared_entry,
+                     converged);
```

Comment:
`beta` unlike `alpha` is used only once. Do we need to store it separately ? Or can it just be computed implicitly in the `update_p` function ?

Reply:
Now it is computed implicitly in the update_p function.